File: src/main/java/com/github/jaksonlin/testcraft/core/annotations/AnnotationParser.java
package com.github.jaksonlin.testcraft.core.annotations;

import com.github.jaksonlin.testcraft.core.context.UnittestCase;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class AnnotationParser {
    private final AnnotationSchema schema;
    private final AnnotationValidator validator;

    public AnnotationParser(AnnotationSchema schema) {
        this.schema = schema;
        this.validator = new AnnotationValidator(schema);
    }

    public UnittestCase parseAnnotation(Map<String, Object> annotationValues) {
        AnnotationValidator.ValidationResult result = validator.validate(annotationValues);

        if (result instanceof AnnotationValidator.ValidationResult.Valid) {
            Map<String, Object> parsedValues = new HashMap<>();
            for (AnnotationFieldConfig field : schema.getFields()) {
                if (field.isRequired()) {
                    if (!annotationValues.containsKey(field.getName())) {
                        throw new IllegalArgumentException("Missing required field: " + field.getName());
                    }
                    if (annotationValues.get(field.getName()) == null) {
                        throw new IllegalArgumentException("Required field cannot be null: " + field.getName());
                    }
                } else {
                    if (!annotationValues.containsKey(field.getName()) || annotationValues.get(field.getName()) == null) {
                        parsedValues.put(field.getName(), field.getDefaultValue());
                        continue;
                    }
                }
                Object rawValue = annotationValues.get(field.getName());
                parsedValues.put(field.getName(), convertValue(rawValue, field));
            }
            return new UnittestCase(parsedValues);
        } else if (result instanceof AnnotationValidator.ValidationResult.Invalid) {
            AnnotationValidator.ValidationResult.Invalid invalidResult = (AnnotationValidator.ValidationResult.Invalid) result;
            throw new IllegalArgumentException(
                    "Invalid annotation values:\n" + String.join("\n", invalidResult.getErrors())
            );
        }
        // Should not happen if ValidationResult is properly implemented
        throw new IllegalStateException("Unexpected ValidationResult type");
    }

    @Nullable
    private Object convertValue(@Nullable Object value, AnnotationFieldConfig field) {
        if (value == null) {
            DefaultValue defaultValue = field.getDefaultValue();
            if (defaultValue instanceof DefaultValue.StringValue) {
                return ((DefaultValue.StringValue) defaultValue).getValue();
            } else if (defaultValue instanceof DefaultValue.StringListValue) {
                return ((DefaultValue.StringListValue) defaultValue).getValue();
            } else if (defaultValue instanceof DefaultValue.NullValue) {
                return null;
            }
            return null; // Should ideally not reach here if DefaultValue is exhaustive
        }

        switch (field.getType()) {
            case STRING:
                return value instanceof String ? value : (field.getDefaultValue() instanceof DefaultValue.StringValue ? ((DefaultValue.StringValue) field.getDefaultValue()).getValue() : null);
            case STRING_LIST:
                if (value instanceof List<?>) {
                    return ((List<?>) value).stream()
                            .filter(String.class::isInstance)
                            .map(String.class::cast)
                            .collect(Collectors.toList());
                } else {
                    return java.util.Collections.emptyList();
                }
            default:
                return null; // Should not happen if AnnotationFieldType is exhaustive
        }
    }
}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/annotations/AnnotationParser.kt
package com.github.jaksonlin.pitestintellij.annotations

import com.github.jaksonlin.pitestintellij.context.UnittestCase


class AnnotationParser(private val schema: AnnotationSchema) {
    private val validator = AnnotationValidator(schema)

    fun parseAnnotation(annotationValues: Map<String, Any>): UnittestCase {
        when (val result = validator.validate(annotationValues)) {
            is AnnotationValidator.ValidationResult.Valid -> {
                val parsedValues = schema.fields.associate { field ->
                    if (field.required) {
                        if (!annotationValues.containsKey(field.name)) {
                            throw IllegalArgumentException("Missing required field: ${field.name}")
                        }
                        if (annotationValues[field.name] == null) {
                            throw IllegalArgumentException("Required field cannot be null: ${field.name}")
                        }
                    } else {
                        if (!annotationValues.containsKey(field.name) || annotationValues[field.name] == null) {
                            return@associate field.name to field.defaultValue
                        }
                    }
                    val rawValue = annotationValues[field.name]
                    field.name to convertValue(rawValue, field)
                }
                return UnittestCase(parsedValues)
            }
            is AnnotationValidator.ValidationResult.Invalid -> {
                throw IllegalArgumentException(
                    "Invalid annotation values:\n${result.errors.joinToString("\n")}"
                )
            }
        }
    }

    private fun convertValue(value: Any?, field: AnnotationFieldConfig): Any? {
        if (value == null) {
            return when (val defaultValue = field.defaultValue) {
                is DefaultValue.StringValue -> defaultValue.value
                is DefaultValue.StringListValue -> defaultValue.value
                DefaultValue.NullValue -> null
            }
        }

        return when (field.type) {
            AnnotationFieldType.STRING -> value as? String ?: field.defaultValue
            AnnotationFieldType.STRING_LIST -> (value as? List<*>)?.mapNotNull { it as? String } ?: emptyList<String>()
            AnnotationFieldType.STATUS -> (value as? String)?.uppercase() ?: field.defaultValue
        }
    }
}